tests/spicy/types/unit/sync-advance.spicy

```diff
@@ -0,0 +1,75 @@
+# @TEST-EXEC: spicyc -dj -o test.hlto %INPUT
+# @TEST-EXEC: spicyc -c %INPUT | (grep '__feat%Test@@X1%sync_advance'; true) >output
+# @TEST-EXEC: spicy-driver -F trace.dat test.hlto >>output
+# @TEST-EXEC: btest-diff output
+#
+# @TEST-DOC: Exercises %sync_advance.
+
+# Case 1: Inline %sync_advance
+
+module Test;
+
+public type X1 = unit {
+    %port = 80/tcp;
+
+    xs: (/(A|B|C)/ &synchronize)[];
+
+    on %synced {
+        confirm;
+    }
+
+    on %sync_advance(offset: uint64) {
```

Comment:
Attaching general feedback here.

Since that the primary use case for this hook would be to raise a parse error if we do not synchronize fast enough, having this hook for every advance in sync mode in general seems like an expensive to me (even if we advanced less often by implementing #1133). That we do not emit it if it is unused does not help parsers which want to handle such scenarios. The original idea with providing a `%gap` hook might have been insufficient, but at least it likely would have led to us invoking the hook orders of magnitude less.

To get a feel for the cost of invoking this hook I ran the following parser:

```ruby
module foo;

public type X = unit {
    : (b"1" &synchronize)[];

    # Leave this in or comment it out to compare hook cost.
    # Also call `$stream.statistics()` in hook for 3rd benchmark.
    on %sync_advance(offset: uint64) { }

    on %synced {
        confirm;
    }
};
```

I ran this on an input file containing containing a `1` followed by 10,000,000 `0`, and then another `1`. I got the following timings:

```console
$ hyperfine -L h wo_hook,w_hook,w_hook_statistics 'spicy-driver {h}.hlto -f data'
Benchmark 1: spicy-driver wo_hook.hlto -f data
  Time (mean ± σ):     994.3 ms ±  17.2 ms    [User: 981.3 ms, System: 11.5 ms]
  Range (min … max):   960.9 ms … 1028.4 ms    10 runs

Benchmark 2: spicy-driver w_hook.hlto -f data
  Time (mean ± σ):      1.291 s ±  0.012 s    [User: 1.276 s, System: 0.013 s]
  Range (min … max):    1.276 s …  1.320 s    10 runs

Benchmark 3: spicy-driver w_hook_statistics.hlto -f data
  Time (mean ± σ):      1.399 s ±  0.019 s    [User: 1.382 s, System: 0.015 s]
  Range (min … max):    1.376 s …  1.430 s    10 runs

Summary
  spicy-driver wo_hook.hlto -f data ran
    1.30 ± 0.03 times faster than spicy-driver w_hook.hlto -f data
    1.41 ± 0.03 times faster than spicy-driver w_hook_statistics.hlto -f data
```

I.e. just having the hook invoked increased the parsing cost by 30%, and doing a minimal operation (just accesing `$stream.statistics()`) adds another 10%. This looks like a lot.

I am wondering whether we need a way to control how often such a hook gets invoked. We could introduce a notion of "scheduled" hook which gets triggered after a certain amount of data has passed through the parser, e.g.,

```ruby
public type X = unit {
   # ...
    on %init {
        # When entering search mode invoke `%sync_mode_scheduled` every 10,000 bytes.
        # Horrible name.
        spicy::sync_mode_schedule_every(10000);
    }

    # This is a hook so it has access to `self`.
    # Also horrible name.
    on %sync_mode_scheduled(offset: uint64) { $stream.statistcs(); }
};
```

I am not sure such a scheduled hook would be useful outside of sync mode.

Reply:
Thanks, for the measurements, that's pretty bad. I was trying to do this without the need for configuration (like you do with the scheduling) nor additional runtime state. But maybe that's not viable. 

Not sure I like the dynamic schedule because that value would need to be stored inside the unit and then retrieved for each check. How about a compile-property property instead with some reasonable default:

```
public type X = unit {
    %sync_advance_block_size = 32768; # if absent, default to some reasonable value, like 4K

    # Executes during synchronization whenever the current offset 
    # crosses a `%sync_advance_block_size` boundary
    on %sync_advance_block(offset: uint64) { $stream.statistcs(); }
};
```

(Open to better names).

Reply:
I like the idea with a property since not a lot of this is dynamic anyway (properties can already take an `expr` so the exact value can be made easily customizable via e.g., Zeek config values). The only downside is that one cannot set property values outside of the unit like one can add hooks, but that is probably okay.

> `    %sync_advance_block_size = 32768; # if absent, default to some reasonable value, like 4K`

I would even go so far as requiring both property and hook if either is specified. We could set some default, but this is a very low-level tool and it might be worth forcing users to think about what they want.

Reply:
>  The only downside is that one cannot set property values outside of the unit like one can add hooks, but that is probably okay.

Not quite like hooks but one can define it at the module level to set a different global default.

> I would even go so far as requiring both property and hook if either is specified. We could set some default, but this is a very low-level tool and it might be worth forcing users to think about what they want.

I think I prefer having a default. Assuming this is relevant only for large input stream anyways, the concrete value doesn't even matter that much. So we'd pick reasonable default (as in: unlikely to hurt performance) but allow fine-tuning.
 
> I think your suggestions work well together and make sense. The only potentially confusing part could be `block` since this is not about _chunk_, WDYT about `%sync_advance_check_interval` and `sync_advance_check`?

I don't like "interval" because it suggests something time-based to me (especially because we also have an `interval` type). Same goes for the earlier "schedule". I picked `block` so that's it different from "chunk", but yeah, I admit that subtle.